proposals/1859_github-auth/README.md

```diff
@@ -0,0 +1,527 @@
+  access to Kubernetes clusters.
+- Regardless of which flow the user uses to authenticate, regularly check with GitHub to see if the user's group
+  memberships have changed and update them accordingly.
+- Any client of the Supervisor should be able to authenticate GitHub users using the web browser-based
+  authentication flow, including the Pinniped CLI and third-party clients registered with the Supervisor as OIDCClients.
+- The IDP chooser page and the IDP discovery endpoints should both include the configured GitHub identity provider(s)
+  for the FederationDomain in the lists that they return.
+- Avoid requiring any changes to the Pinniped CLI or other clients. From a client's point of view,
+  which external identity provider is used for authentication should be transparent.
+
+### Specification / How it Solves the Use Cases
+
+#### New GitHubIdentityProvider resource
+
+Add a new CRD called GitHubIdentityProvider. Here is an example:
+
+```yaml
+apiVersion: idp.supervisor.pinniped.dev/v1alpha1
+kind: GitHubIdentityProvider
+metadata:
+  namespace: supervisor
+  name: github
+spec:
+  github_api:
+    # Required only for GitHub Enterprise Server (on-prem).
+    # Defaults to using the GitHub's public API.
+    host: my.example.com
+    # X.509 Certificate Authority (base64-encoded PEM bundle).
+    # If omitted, a default set of system roots will be trusted.
+    # Required only for GitHub Enterprise Server (on-prem), and only
+    # when its CA will not be trusted by the default system roots.
+    certificateAuthorityData: LS0tLS1CRUdJTiBDRV...0tLS0tCg==
+  claims:
+    # Which property of the GitHub user record shall determine
+    # the username in Kubernetes. Can be either "id", "login",
+    # or "login:id".
+    # Login names can only contain alphanumeric characters and
+    # non-repeating hyphens, and may not start or end with hyphens.
+    # GitHub's users are allowed to change their login name,
+    # although it is inconvenient. If a GitHub user changed
+    # their login name from "foo" to "bar", then a second user
+    # might change their name from "baz" to "foo" in order
+    # to take the old username of the first user. For this
+    # reason, it is not as safe to make authorization decisions
+    # based only on the login name.
+    # If desired, an admin could configure identity
+    # transformation expressions on a FederationDomain to
+    # further customize these usernames.
+    # Defaults to "login:id", which is the login name, followed
+    # by a colon, followed by the unique and unchanging integer
+    # ID number. This blends human-readable login names with the
+    # unchanging ID value. Note that colons are not allowed in
+    # GitHub login names or in ID numbers, so the colon in the
+    # "login:id" name will always be the login/id separator colon.
+    username: id
+    # Which property of the GitHub team record shall determine
```

Comment:
We should perhaps clarify in documentation that the list of groups will only be the groups to which the user directly belongs as provided by the GH API. Parent teams are not included. The attached example would result in team membership of `team1`, `team3`, and `:iamateam` by name, which have slugs `team1`, `team3`, and `iamateam`, respectively.
![image](https://github.com/vmware-tanzu/pinniped/assets/24278092/4adc98ba-80d2-4009-9243-e14c128304b8)

Reply:
If the end-user (the Pinniped admin) desires to see all of the user's nested group membership, this would have to be done by additional API calls from Pinniped to unroll every group's parent membership. This could result in a a number of API calls that are **_orders of magnitude greater_** than expected, depending on the number and complexity of the organization's team structure. There is no limit in GH to the nesting level or number of teams that exist at any given level within the organization.

Reply:
It's possible we could defer any features about exposing nested group membership to after the initial release of a `GitHubIdentityProvider` and prioritize that based on community feedback and requests for nested groups.

Reply:
We could investigate using the GraphQL API as an easier way to access the nested groups that a user belongs to, but not that this API is also rate-limited and would require investigation.

Reply:
> We should perhaps clarify in documentation that the list of groups will only be the groups to which the user directly belongs as provided by the GH API. Parent teams are not included. 

> The API does not flatten teams. It only returns the teams for which you are a direct member, and for any nested team for which you are a direct member it also says which team is the immediate parent of that team. So we could either only report your direct team memberships, or we would also include the first level of hierarchy above your direct team memberships, but we cannot include the whole hierarchy above your direct team memberships without making more API calls (potentially lots of API calls which cut into your hourly rate limit).

TODO: Add this to the proposal. We will show your direct team memberships as well as the direct parents of those teams. We will be careful to make this a unique set (i.e. remove duplicates).